blues/sparrow-reference-web-app
# feat: sparrow signal strength

# Problem Context

As we roll towards the May 6th cutoff date, gateways report their signal strength as "bars". Sparrow nodes report their signal strength and other health metrics to the sensors.db notefile. Each gateway and sensor block in the app should display an indication of the signal strength.

Gateways show the signal strength of their cellular or wifi connection as 0-4 bars.

Nodes show the combined signal strength of the LoRa transmission from the node and the gateway. (RSSI/SNR for node and gateway, normalized to 0-4 bars.)

I implemented the React components to display the appropriate signal strength values on either gateways or nodes throughout the app according to the design.

## Changes

Added all the new SVG icons for the various wifi, cell and lora signal strength options.
Made 3 new helper functions to figure out which signal strength image to display in various parts of the UI.
Added the icons to the appropriate pages to match the mockups provided by design.
Added the new properties to the necessary gateway and node models and their subsequent view models. 

## Screenshot (if applicable)

Main page (cell bars and lora bars)

![Screen Shot 2022-04-22 at 10 04 08 AM](https://user-images.githubusercontent.com/20400845/164731545-a84d98e4-00e5-4c0c-8477-c7307076cf03.png)

Gateway page (wifi bars)

![Screen Shot 2022-04-22 at 10 03 16 AM](https://user-images.githubusercontent.com/20400845/164731578-7f4165b7-e09d-4085-9448-27aa4e4ac9ea.png)

Node page (lora bars)

![Screen Shot 2022-04-22 at 10 06 24 AM](https://user-images.githubusercontent.com/20400845/164731608-c737ac9b-d6c3-4da0-a40b-6aeb83b1d696.png)

## Testing

Unit / integration / e2e tests?

Fixed up unit tests, didn't bother with e2es at this stage in the game.

Steps to test manually?

1. Start up the app and go to http://localhost:4000
2. To see homepage display signal strength, on the `GatewayCard.tsx` file, add the following around line 35 of the file: `gatewayDetails.cellBars = "2"` or `gatewayDetails.wifiBars = "3"`
3. To see a gateway page display signal strength, on the `GatewayDetails.tsx` file, add a `return` statement to the component, then add `viewModel.gateway.cellBars = "0"` or `viewModel.gateway.wifiBars = "4"`, around line 26 - 27
4. To see a node page display signal strength, on the `NodeCard.tsx` file, add the following line around line 29 of the file: `nodeDetails.bars = 1`

Any other sorts of testing notes to include?

## Any other related PRs

n/a

## Ticket(s)

https://www.pivotaltracker.com/story/show/181871880
